nexus/src/app/quiesce.rs

```diff
@@ -112,31 +141,76 @@ impl NexusQuiesceHandle {
     }
 }
 
-async fn do_quiesce(quiesce: NexusQuiesceHandle) {
-    let saga_quiesce = quiesce.sagas.clone();
+async fn do_quiesce(mut quiesce: NexusQuiesceHandle) {
     let datastore = quiesce.datastore.clone();
+    let my_nexus_id = quiesce.my_nexus_id;
+
+    /// minimal timeout used just to avoid tight loops when encountering
+    /// transient errors
+    const PAUSE_TIMEOUT: Duration = Duration::from_secs(5);
 
-    // NOTE: This sequence will change as we implement RFD 588.
-    // We will need to use the datastore to report our saga drain status and
-    // also to see when other Nexus instances have finished draining their
-    // sagas.  For now, this implementation begins quiescing its database as
-    // soon as its sagas are locally drained.
     assert_matches!(
         *quiesce.state.borrow(),
         QuiesceState::DrainingSagas { .. }
     );
 
-    // TODO per RFD 588, this is where we will enter a loop, pausing either on
-    // timeout or when our local quiesce state changes.  At each pause: if we
-    // need to update our db_metadata_nexus record, do so.  Then load the
-    // current blueprint and check the records for all nexus instances.  This
-    // work is covered by a combination of oxidecomputer/omicron#8859,
-    // oxidecomputer/omicron#8857, and oxidecomputer/omicron#8796.
+    // We've recorded that we should be quiescing.  This will prevent new sagas
+    // from starting.  However, we can still wind up running new sagas if a
+    // blueprint execution re-assigns us saga from an expunged Nexus.
+    //
+    // So here's the plan: we keep track of the blueprint id as of which we have
+    // fully drained.  That means that we've processed all Nexus expungements up
+    // to that blueprint, _and_ we've recovered all sagas assigned to us, _and
```

Comment:
Nit - missing the closing `_` for the `_and` on the end of this line

Reply:
Fixed in e8eb204f96309d21e84e43feee2b09ad635d7ddd.